### DISCUSSION TITLE
IDOM Stalling Due to Rapid Refreshes

### DISCUSSION THREAD

**Discussion by Archmonger** [idom-team/idom#524] [Problem] [answered]
Originates from https://github.com/idom-team/django-idom/issues/22

## Summary
IDOM can stall during some conditions (no longer capable of rendering new components), and the only solution is waiting it out. 
This fully breaks IDOM for web development, since IDOM completely halts whenever a webpage is refreshed too quickly.

## Steps to Reproduce
Rapidly refresh the client (webpage in a Chromium browser), or rapidly refresh the backend (ex. via Django Autoreload).

## Analysis and Details
- Seems to be able to be replicated on any ASGI websocket
- The client can get stalled if a client side WS disconnection event doesn't occur (until WS keep-alive timeout)

## OLD: Analysis and Details
- Perhaps this is related to refreshing the page in the middle of the element's first render.
- My best guess is this is some bug related to `dispatch_single_view`, which means it digs into IDOM Core.
- I've confirmed I can also completely break IDOM [on our server used for IDOM docs examples](https://idom-docs.herokuapp.com/docs/index.html).
- Also occurs [Conreq's app store branch](https://github.com/Archmonger/Conreq/pull/69) on Windows 10
- Somehow restarting the python interpreter does not resolve this.
- Issue automagically fixes itself after waiting 1-2 minutes. Something is timing out but I can't determine what.

**Comment by rmorshea** [comment, marked as answer]
I'm able to reproduce when running the docs locally. This is a pretty bad bug so it'll be top priority unless anything else comes up.

**Reply by Archmonger** [reply]
Agreed, if there's anything I can do to support let me know.

**Reply by rmorshea** [reply]
Next bit of evidence...

*the plot thickens*

After hitting refresh a number of times the server simply stops accepting new web socket connections. Then, after some time, a number of `KeepAlive Timeout. Closing connection.` debug messages get logged in the server.

Best guess for now is that there's something wrong with the client and that it's not properly releasing its connection to the server?

**Reply by rmorshea** [reply]
And, after a quick google search that looks like a probable source of this problem - turns out web sockets don't automatically disconnect when the page unloads: https://stackoverflow.com/a/65699403/3159288

Who would have thought 🤷

**Reply by rmorshea** [reply]
It's weird that new web socket connections aren't getting rejected though...

**Reply by Archmonger** [reply]
I've confirmed I can replicate this on a [simple websocket](https://github.com/Archmonger/Conreq/blob/a11f5748e9a41646bcce35fb70c259c6cac61905/conreq/internal/websockets/consumers.py#L11) with no IDOM.

**Reply by Archmonger** [reply]
I confirmed this only affects a single client.

The webserver is willing to accept new connections from any other client. 

As you suggested, the solution is going to be a window unload event.

**Reply by rmorshea** [reply]
Yup. Here's a gif demonstrating this
![bug1](https://user-images.githubusercontent.com/4041990/139217863-0df8d0f7-9ca5-4a42-8720-e48873ca47a5.gif)
:

**Reply by Archmonger** [reply]
Well this is really good news. Thankfully this didn't turn into some mysterious backend bug without a stack trace.

**Reply by rmorshea** [reply]
Ok, so the bug's not so bad since it's isolated to a particular client. But it's still pretty bad since this issue can persist across tabs.

Interestingly, in Firefox (seen left), while it's still possible to trigger this issue (though it's harder) pausing briefly, and doing just one more refresh, always clears it up. I guess Firefox must do a more thorough job of clearing up resources from the page on each refresh than Chromium-based browsers (seen right).

**Reply by Archmonger** [reply]
EDIT: Only didn't work in Brave

I have some bad news.

I attempted adding an unload listener on the simple Django websocket example linked above.

It did not resolve the issue.
```js
window.addEventListener("unload", function () {
	if (COMMAND_SOCKET.readyState == WebSocket.OPEN) COMMAND_SOCKET.close();
});
```

Perhaps we also need to be looking for states other than just OPEN. I'll continue to test and report back.

**Reply by Archmonger** [reply]
The solution above seems to work in all the chromium variants besides Brave. Additionally, issue no longer occurs on FireFox. 

I have no idea why Brave doesn't like this.

**Reply by rmorshea** [reply]
does it work if we just do `COMMAND_SOCKET.close();` without a state check?

**Reply by rmorshea** [reply]
Also for reference when we implement this. The relevant code for closing seems to be `1001` "going away".

**Reply by Archmonger** [reply]
These also don't work in Brave

```javascript
window.addEventListener("unload", function () {
	COMMAND_SOCKET.close();
});

```
```javascript
window.addEventListener("unload", function () {
	try {
		COMMAND_SOCKET.close();
	} catch (error) {
		// Nothing is ever printed
		console.log(error);
		return;
	}
});
```

**Reply by Archmonger** [reply]
Since it's only Brave out of the Chromium bunch that has this issue, I would suspect it's something specific to Brave that we can't resolve.

**Reply by Archmonger** [reply]
GH issue for brave has been created https://github.com/brave/brave-browser/issues/19080

**Reply by rmorshea** [reply]
It looks like there's a [`beforeunload`](https://developer.mozilla.org/en-US/docs/Web/API/Window/beforeunload_event) that might work too?

**Reply by Archmonger** [reply]
Tested that as well last night, didn't change the behavior.

**Comment by Archmonger** [comment]
https://github.com/brave/brave-browser/issues/19080 <- My report
https://github.com/brave/brave-browser/issues/19123
https://github.com/brave/brave-browser/issues/15410

Seems like similar flavors of this issue have been reported before on the Brave side. So we can safely ignore Brave's oddball behavior for now and add an `unload` event handler for every other browser.
